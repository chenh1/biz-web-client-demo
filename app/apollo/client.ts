// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});