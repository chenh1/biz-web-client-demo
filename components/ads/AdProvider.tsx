"use client"
import type { AdTags } from "./tags/types"

import { usePathname } from "next/navigation"
import Script from "next/script"
import React, { useEffect, useState, createContext, useRef, Context } from "react"
import { Overlay } from "../containers/Overlay/Overlay"
import { Heading } from "../core/Heading/Heading"
import { Paragraph } from "../core/Paragraph/Paragraph"
import { Box } from "../containers/Box/Box"
import { useFlippers } from "../utils/useFlippers"

const NETWORK_FAILS_THRESHOLD: number = 2
const COLLAPSED_ELEMENT_CHECKS_THRESHOLD: number = 6

export const AdContext: Context<{ adScriptsLoaded: boolean; shouldShow: boolean; }> = createContext({ adScriptsLoaded: false, shouldShow: false } as { adScriptsLoaded: boolean; shouldShow: boolean; })

export interface ProviderContextType {
  adScriptsLoaded: boolean,
  shouldShow: boolean
}

export type AdProviderType = {
  tags: AdTags,
  children?: string | JSX.Element | JSX.Element[], 
}

export const AdProvider: React.FC<AdProviderType> = ({ tags, children }: AdProviderType) => {
  const networkCheckInterval = useRef<NodeJS.Timeout | undefined>(undefined)
  const collapsedElementCheckInterval = useRef<NodeJS.Timeout | undefined>(undefined)
  const pathName = usePathname();
  const [ adScriptsLoaded, setAdScriptsLoaded ]: [boolean, (boolean) => void] = useState(false)
  const pathChanged = useRef(undefined)
  const { ads: shouldShow } = useFlippers({ key: 'ads' });
  

  const adChecks = useRef({ networkFails: 0, collapsedElementChecks: 0 })
  const [ adChecksState, setAdChecksState ] = useState(adChecks.current)

  const updateAdChecksState = (newAdChecks) => {
    adChecks.current = newAdChecks
    setAdChecksState(newAdChecks)
  }
  
  
  useEffect(() => {
    if (typeof window !== 'undefined' && shouldShow) {
      window.googletag = window.googletag || { cmd: [] };

      window.googletag.cmd.push(function() {
        window.googletag.destroySlots?.();

        tags?.forEach(({ elId, path, isResponsive, slots, isIterative }) => {
          if (isResponsive && !adScriptsLoaded) {
            // check if responsive ad slot has already been defined
            if (!isIterative || (isIterative && document.getElementById(elId))) {
              const responsiveAdSlot = window.googletag.defineSlot?.(path, slots, elId).addService(window.googletag.pubads?.());
            
              const mapping = window.googletag.sizeMapping?.()
                .addSize([1024, 768], [[750, 200], [728, 90]])
                .addSize([640, 480], [300, 250])
                .addSize([0, 0], [])
                .build();
              
              responsiveAdSlot?.defineSizeMapping(mapping);
            }
          } else {
            window.googletag
              .defineSlot?.(path, slots, elId)
              .addService(window.googletag.pubads?.());
          }
        })

        window.googletag.enableServices?.();
      })
      
      setAdScriptsLoaded(true)
    }
  }, [ pathName, shouldShow ])

  useEffect(() => {
    if (typeof window !== 'undefined' && shouldShow) {
      networkCheckInterval.current = setInterval(() => {
        fetch(`https://www.googletagmanager.com/gtm.js?id=${process.env.GTM_ID}`)
          .then((res) => {
            clearInterval(networkCheckInterval.current)
          })
          .catch((e) => {
            updateAdChecksState({
              ...adChecks.current,
              networkFails: adChecks.current.networkFails + 1
            })

            if (adChecks.current.networkFails >= NETWORK_FAILS_THRESHOLD) {
              clearInterval(networkCheckInterval.current)
            }
          })
      }, 2000)

      collapsedElementCheckInterval.current = setInterval(() => {
        const adContainer = document.querySelector('.js-ad-container')
        if (!!adContainer && typeof adContainer?.children[0] === 'undefined') {
          updateAdChecksState({
            ...adChecks.current,
            collapsedElementChecks: adChecks.current.collapsedElementChecks + 1
          })
          
          if (adChecks.current.collapsedElementChecks >= COLLAPSED_ELEMENT_CHECKS_THRESHOLD) {
            clearInterval(collapsedElementCheckInterval.current)
          }
        }
      }, 1500)
    }

  }, [])
  
  const shouldShowAdWarningOverlay = adChecks?.current?.networkFails >= NETWORK_FAILS_THRESHOLD && adChecks?.current?.collapsedElementChecks >= COLLAPSED_ELEMENT_CHECKS_THRESHOLD
  
  return (
    <>
      <Script async src="https://securepubads.g.doubleclick.net/tag/js/gpt.js" />
      <AdContext.Provider value={{ adScriptsLoaded, shouldShow }}>
        <Overlay toggleVisibility={() => {}} visibility={shouldShowAdWarningOverlay}>
          <Box gap="sm" className="max-w-[1080px]">
            <Heading type="h3">Hey there, about your ad blocker...</Heading>
            <Paragraph>
              We get it. Ads can be annoying. We're offering you content for free and we have server bills to pay for to keep this up and running. Please take this into consideration. If you're cool with it, please turn it off and refresh the page. If you feel like you absolutely cannot turn off your ad blocker, then please navigate away from this website. Thanks for your time; we appreciate it!
            </Paragraph>
          </Box>
        </Overlay>
        {children}
      </AdContext.Provider>
    </>
  )
}