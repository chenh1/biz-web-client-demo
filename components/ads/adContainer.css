.ad-wrapper.center > div:first-of-type {
  display: grid;
  justify-content: center;
}

.ad-wrapper.left > div:first-of-type {
  display: grid;
  justify-content: start;
}

.ad-wrapper.right > div:first-of-type {
  display: grid;
  justify-content: end;
}