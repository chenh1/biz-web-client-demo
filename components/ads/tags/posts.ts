import type { AdTags } from "./types"

const adTags: AdTags = [
  // {
  //   elId: "single-fixed-ad",
  //   path: '/6355419/Travel/Europe/France/Paris',
  //   isResponsive: false,
  //   slots: [300, 250]
  // },
  // {
  //   elId: "base-ad",
  //   path: '/6355419/Travel/Europe',
  //   isResponsive: false,
  //   slots: [[300, 250], [728, 90], [750, 200]]
  // },
  // {
  //   elId: "banner-ad",
  //   path: '/6355419/Travel/Europe',
  //   isResponsive: false,
  //   slots: [[300, 250], [728, 90], [750, 200]]
  // },
  {
    elId: "responsive-ad-1",
    path: '/6355419/Travel/Europe',
    isResponsive: true,
    slots: [[300, 250], [728, 90], [750, 200]]
  },
  {
    elId: "responsive-ad-2",
    path: '/6355419/Travel/Europe',
    isResponsive: true,
    slots: [[300, 250], [728, 90], [750, 200]]
  },
]

export default adTags