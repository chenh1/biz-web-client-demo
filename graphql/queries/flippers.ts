import { gql, DocumentNode } from "@apollo/client";

export const flippersQuery: DocumentNode = gql`
  query FlipperManager {
    flipperManager {
      data {
        attributes {
          ads
          adsPostsPage
          adsPostPage
          adsDashboard
          adsModuleLearnView
          adsModuleExamView
          testMode
          maintenanceMode
        }
      }
    }
  }
`;
