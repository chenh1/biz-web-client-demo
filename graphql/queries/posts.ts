import { gql, DocumentNode } from "@apollo/client";
import type { PostType } from "./post";

export type PostsRes = {
  posts: {
    data: PostType[]
  }
}

export const postsQuery: DocumentNode = gql`
  query Posts($page: Int, $pageSize: Int, $in: [String], $sort: [String], $ne: ID, $tagNe: String) {
    posts (sort: $sort, pagination: { page: $page, pageSize: $pageSize }, filters: { id: { ne: $ne }, tags: {value: { in: $in, ne: $tagNe }} } ) {
      data {
        id
        attributes {
          slug
          previewImage {
            data {
              attributes {
                url
                alternativeText
                width
                height
              }
            }
          }
          tags {
            data {
              attributes {
                displayName
                value
                color
              }
            }
          }
          title
          createdAt
          publishedAt
          author
        }
      }
    }
  }
`