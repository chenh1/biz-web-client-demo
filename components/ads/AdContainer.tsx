"use client"

import React, { useEffect, useContext } from "react";
import { AdContext } from "./AdProvider";
import './adContainer.css';

type AdContainerType = {
  elId: string;
  align: "center" | undefined;
}

export const AdContainer: React.FC<AdContainerType> = ({ elId = "base-ad", align = "center" }) => {
  const { adScriptsLoaded, shouldShow } = useContext<{ adScriptsLoaded: boolean; shouldShow: boolean; }>(AdContext)
  
  useEffect(() => {
    if (typeof window !== 'undefined' && adScriptsLoaded && shouldShow) {
      window.googletag.cmd.push(function() {
        window.googletag.display?.(elId);
      });
    }
  }, [ adScriptsLoaded ])

  return shouldShow
    ? <div className={`ad-wrapper ${align}`}>
        <div className="js-ad-container" id={elId}></div>
      </div>
    : null
}