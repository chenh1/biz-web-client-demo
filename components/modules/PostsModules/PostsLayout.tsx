'use client'

import type { PostType } from "@/graphql/queries/post";
import React, { useState } from "react"
import { useQuery } from "@apollo/client";
import { Box } from "../../containers/Box/Box";
import { postsQuery } from "../../../graphql/queries/posts";
import Link from "next/link";
import { Heading } from "../../core/Heading/Heading";
import { Pill } from "../../core/Pill/Pill";
import { Divider } from "../../core/Divider/Divider";
import { Hover3dCard } from "../../animations/Hover3dCard/Hover3dCard";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import { PostsList } from "./PostsList";
//import { UnderMaintenanceCatcher } from "../UnderMaintenanceCatcher/UnderMaintenanceCatcher";
import { AdContainer } from "../../ads/AdContainer";
import { useFlippers } from "../../utils/useFlippers";
import { Loader } from "../Loader/Loader";

export const PostsLayout = () => {
  const [ page, setPage ] = useState<number>(1);
  const [ pageSize, setPageSize ] = useState<number>(12);
  const { testMode } = useFlippers({ key: 'testMode' });

  const { data: { posts: { data } = {} } = {}, loading: postsLoading, error: postsError, fetchMore } = useQuery(postsQuery, {
    variables: {
      page: page,
      pageSize: pageSize,
      sort: [ "createdAt:desc" ]
    }
  });

  const getNextPage = () => {
    const nextPage = page + 1;
    setPage(nextPage);
    fetchMore({
      variables: {
        page: nextPage,
        pageSize: pageSize,
        sort: [ "createdAt:desc" ]
      }
    })
  }

  const posts: PostType[] = data?.filter(({ attributes: { tags } }) => {
    if (!testMode) {
      const isTestArticle = tags?.data?.some(({ attributes: { value } }) => value === 'testArticle')
      return !isTestArticle
    } else {
      return true
    }
  }) || []

  const featuredPost = posts?.[0]?.attributes
  const featuredPreviewImage = featuredPost?.previewImage?.data?.attributes;
  const { tags: featuredPostTags } = featuredPost || {};

  const secondaryPosts: PostType[] = posts?.slice(1, 5)
  const remainingPosts: PostType[] = posts?.slice(5)

  return (
    // <UnderMaintenanceCatcher error={(postsError || posts?.length === 0) && !postsLoading} noData>
    <>
      <AdContainer align="center" elId="responsive-ad-1"/>
      <Heading type="h1">Latest Posts</Heading>
      {postsLoading ?
        <Box padding="0" gap="md">
          <Loader/>
          <Heading className="text-center" type="h4">Sprouting up the latest headlines...</Heading>
        </Box>
        :
        <>
          {featuredPost &&
            <Link href={`posts/${featuredPost?.slug}`}>
              <Hover3dCard>
                <Box hasShadow isRounded className="grid-cols-1 sm:grid-cols-6 justify-center items-center hover:shadow-xl">
                  <Box padding="0" className="col-span-2">
                    {featuredPreviewImage && <img src={featuredPreviewImage?.url} width={featuredPreviewImage?.width} height={featuredPreviewImage?.height} alt={featuredPreviewImage?.alternativeText}/>}
                  </Box>
                  <Box padding="0" className="sm:col-span-4">
                    <Heading type="h2" className="text-lg font-semibold">{featuredPost?.title}</Heading>
                    <Paragraph size="sm" className="italic">{`Published ${new Date(featuredPost?.createdAt).toDateString()}`}</Paragraph>
                    <div>
                    {featuredPostTags?.data?.map(({ attributes: { value, displayName, color } }, i) => {
                      return (
                        <Pill key={i} color={color} className="text-x mr-2 mb-2">{displayName}</Pill>
                      )
                    })}
                    </div>
                  </Box>
                </Box>
              </Hover3dCard>
            </Link>
          }
          <Box padding="0" className="justify-center">
            <Box padding="0" className="w-[90vw] max-w-[1100px]">
              <Divider/>
            </Box>
          </Box>
          <Box padding="0" gap="sm" className="justify-center sm:grid-cols-4">
            {secondaryPosts?.map(({ id, attributes: { title, slug, previewImage, tags } }) => {
              const { url: imageUrl, width, height, alternativeText } = previewImage?.data?.attributes || {};

              return (
                <Link key={id} href={`posts/${slug}`}>
                  <Hover3dCard>
                    <Box hasShadow isRounded className="flex flex-col justify-center items-center hover:shadow-xl overflow-hidden min-h-[400px]">
                      <Box padding="0" className="-z-10 scale-[4] h-[200px] items-end relative">
                        <div className="absolute top-0 opacity-25 h-full w-full bg-night-black"></div>
                        <img src={imageUrl} width={width} height={height} alt={alternativeText}/>
                      </Box>
                      <Box className="absolute top-0 dark">
                        <Heading type="h3" className="text-lg font-semibold">{title}</Heading>
                      </Box>
                      <Box padding="0" className="items-end">
                        <div>
                        {tags?.data?.map(({ attributes: { value, displayName, color } }, i) => {
                          return (
                            <Pill key={i} color={color} className="text-x mr-2 mb-2">{displayName}</Pill>
                          )
                        })}
                        </div>
                      </Box>
                    </Box>
                  </Hover3dCard>
                </Link>
              )
            })}
          </Box>
          <Box padding="0" className="justify-center">
            <Box padding="0" className="w-[90vw] max-w-[1100px]">
              <Divider/>
            </Box>
          </Box>
          {remainingPosts?.length > 0 && <PostsList initialPosts={remainingPosts}/>}
        </>
      }
      <AdContainer align="center" elId="responsive-ad-2"/>
    </>
    // </UnderMaintenanceCatcher>
  )
}