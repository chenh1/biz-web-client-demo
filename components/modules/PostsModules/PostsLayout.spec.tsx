import { PostsLayout } from "./PostsLayout";
import { MockedProvider } from "@apollo/client/testing";
import { AdProvider } from "../../ads/AdProvider";
import postTags from '../../ads/tags/post'
import { render, waitFor } from '@testing-library/react';
import { postsQuery } from "../../../graphql/queries/posts";
import { JSDOM } from "jsdom"
import { flippersQuery } from "../../../graphql/queries/flippers";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock("next/link", () => ({ url, children }) => <a href={url}>children</a>);

const mocks = [
  {
    request: {
      query: flippersQuery,
    },
    result: {
      data: {
        flipperManager: {
          data: {
            attributes: { testMode: false }
          }
        }
      }
    }
  },
  {
    request: {
      query: postsQuery,
      variables: {
        page: 1,
        pageSize: 12,
        sort: [ "createdAt:desc" ]
      },
    },
    result: {
      data: {
        posts: {
          data: [
            { 
              id: "0",
              attributes: {
                slug: "/post/0",
                tags: {
                  data: [
                    { attributes: { displayName: "Test Article", value: "equipment", clientFacing: true, color: "hyper-red" } }
                  ]
                },
                title: "Test Article",
                content: "Test Article Content",
                createdAt: "2021-08-02T00:00:00Z",
                publishedAt: "2021-08-02T00:00:00Z",
                author: "Test Author",
                previewImage: {
                  data: {
                    attributes: {
                      url: "https://www.example.com/image.jpg",
                      width: 100,
                      height: 100,
                      alternativeText: "Test Image"
                    }
                  }
                }
              }
            },
            { 
              id: "1",
              attributes: {
                slug: "/post/1",
                tags: {
                  data: [
                    { attributes: { displayName: "Test Article 2", value: "recipes", clientFacing: true, color: "hyper-red" } }
                  ]
                },
                title: "Test Article 2",
                content: "Test Article Content",
                createdAt: "2021-08-02T00:00:00Z",
                publishedAt: "2021-08-02T00:00:00Z",
                author: "Test Author",
                previewImage: {
                  data: {
                    attributes: {
                      url: "https://www.example.com/image.jpg",
                      width: 100,
                      height: 100,
                      alternativeText: "Test Image"
                    }
                  }
                }
              }
            },
          ]
        }
      }
    }
  }
]


it('renders PostsLayout module', async () => {
  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AdProvider tags={postTags}>
        <PostsLayout/>
      </AdProvider>
    </MockedProvider>
  )

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(container).toMatchSnapshot();
});
