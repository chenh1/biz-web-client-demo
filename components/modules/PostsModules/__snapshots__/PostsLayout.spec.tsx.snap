// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PostsLayout module 1`] = `
<div>
  <h1
    class="text-night-black dark:text-max-white font-headings font-semibold text-4xl sm:text-6xl "
  >
    Latest Posts
  </h1>
  <a>
    children
  </a>
  <div
    class=" grid  gap-4  p-2 justify-center"
  >
    <div
      class=" grid  gap-4  p-2 w-[90vw] max-w-[1100px]"
    >
      <hr
        class="w-full h-0 border-0 border-t-[1px] border-night-black-300 dark:border-max-white "
      />
    </div>
  </div>
  <div
    class=" grid  gap-4  p-2 justify-center sm:grid-cols-4"
  >
    <a>
      children
    </a>
  </div>
  <div
    class=" grid  gap-4  p-2 justify-center"
  >
    <div
      class=" grid  gap-4  p-2 w-[90vw] max-w-[1100px]"
    >
      <hr
        class="w-full h-0 border-0 border-t-[1px] border-night-black-300 dark:border-max-white "
      />
    </div>
  </div>
</div>
`;
